ethereum/eip712/encoding.go

```diff
@@ -0,0 +1,207 @@
+package eip712
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/simapp/params"
+	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
+
+	cosmosTypes "github.com/cosmos/cosmos-sdk/types"
+	txTypes "github.com/cosmos/cosmos-sdk/types/tx"
+
+	apitypes "github.com/ethereum/go-ethereum/signer/core/apitypes"
+	ethermint "github.com/evmos/ethermint/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+)
+
+var ethermintProtoCodec codec.ProtoCodecMarshaler
+var ethermintAminoCodec *codec.LegacyAmino
+
+// The process of unmarshaling SignDoc bytes into a SignDoc object requires having a codec
+// populated with all relevant message types. As a result, we must call this method on app
+// initialization with the app's encoding config.
+func SetEncodingConfig(cfg params.EncodingConfig) {
+	ethermintAminoCodec = cfg.Amino
+	ethermintProtoCodec = codec.NewProtoCodec(cfg.InterfaceRegistry)
+}
+
+// Get the EIP-712 object hash for the given SignDoc bytes by first decoding the bytes into
+// an EIP-712 object, then hashing the EIP-712 object to create the bytes to be signed.
+// See https://eips.ethereum.org/EIPS/eip-712 for more.
+func GetEIP712HashForMsg(signDocBytes []byte) ([]byte, error) {
+	var typedData apitypes.TypedData
+
+	// Attempt to decode as both Amino and Protobuf since the message format is unknown.
+	// If either decode works, we can move forward with the corresponding typed data.
+	typedDataAmino, errAmino := decodeAminoSignDoc(signDocBytes)
```

Comment:
Currently if `errAmino == nil` then `decodeProtobufSignDoc` is run unnecessarily

Reply:
Regarding performance, this is relatively cheap. Regarding code-quality, it currently seems better than the alternative:

```
	typedDataAmino, errAmino := decodeAminoSignDoc(signDocBytes)
	if errAmino == nil {
		return typedDataAmino, nil
	}
	typedDataProtobuf, errProtobuf := decodeProtobufSignDoc(signDocBytes)
	if errProtobuf == nil {
		return typedDataProtobuf, nil
	}

	return apitypes.TypedData{}, fmt.Errorf("could not decode sign doc as either Amino or Protobuf.\n amino: %v\n protobuf: %v\n", errAmino, errProtobuf)
```

Either way works for me, let me know what you think.

Reply:
@austinchandra good point, it's fine to keep it.

Reply:
I think its worth modifying this as per @danburck first comment. If we can avoid running an unecessary decoding I think its worth it. As per code quality you can ambstract this into a function of its own so `GetEIP712HashForMsg` stays clean

Reply:
> Regarding code-quality, it currently seems better than the alternative

This alternative looks better than the current implementation. Let's refactor it